engine/src/main/java/org/hibernate/search/engine/impl/ImmutableSearchFactory.java

```diff
@@ -131,15 +131,20 @@ public ImmutableSearchFactory(SearchFactoryState state) {
 		this.timeoutExceptionFactory = state.getDefaultTimeoutExceptionFactory();
 		this.timingSource = state.getTimingSource();
 		this.mapping = state.getProgrammaticMapping();
-		this.statistics = new StatisticsImpl( this );
+		if ( state.getStatistics() == null ) {
+			this.statistics = new StatisticsImpl( this );
+			boolean statsEnabled = ConfigurationParseHelper.getBooleanValue(
+					configurationProperties, Environment.GENERATE_STATS, false
+			);
+			this.statistics.setStatisticsEnabled( statsEnabled );
+		}
+		else {
+			this.statistics = (StatisticsImpl) state.getStatistics();
```

Comment:
right, I have the same question. Would be nice if you could improve the patch to not need the cast, or if that's not possible to add a comment to explain why it is safe.